component/file_cache/file_cache.go

```diff
@@ -780,61 +799,82 @@ func (fc *FileCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Hand
 
-		if attr != nil {
-			// chtimes shall be the last api otherwise calling chmod/chown will update the last change time
-			err = os.Chtimes(localPath, attr.Atime, attr.Mtime)
-			if err != nil {
-				log.Err("FileCache::OpenFile : Failed to change times of file %s [%s]", options.Name, err.Error())
-			}
+	if attr != nil {
+		// chtimes shall be the last api otherwise calling chmod/chown will update the last change time
+		err = os.Chtimes(localPath, attr.Atime, attr.Mtime)
+		if err != nil {
+			log.Err("FileCache::DownloadFile : Failed to change times of file %s [%s]", options.Name, err.Error())
 		}
-
-		fileCacheStatsCollector.UpdateStats(stats_manager.Increment, dlFiles, (int64)(1))
-	} else {
-		log.Debug("FileCache::OpenFile : %s will be served from cache", options.Name)
-		fileCacheStatsCollector.UpdateStats(stats_manager.Increment, cacheServed, (int64)(1))
 	}
 
+	fileCacheStatsCollector.UpdateStats(stats_manager.Increment, dlFiles, (int64)(1))
+
 	// Open the file and grab a shared lock to prevent deletion by the cache policy.
-	f, err = common.OpenFile(localPath, options.Flags, options.Mode)
+	f, err = common.OpenFile(localPath, flag, fMode)
 	if err != nil {
-		log.Err("FileCache::OpenFile : error opening cached file %s [%s]", options.Name, err.Error())
+		log.Err("FileCache::DownloadFile : error opening cached file %s [%s]", options.Name, err.Error())
 		return nil, err
 	}
 
 	// Increment the handle count in this lock item as there is one handle open for this now
 	flock.Inc()
 
-	handle := handlemap.NewHandle(options.Name)
 	inf, err := f.Stat()
 	if err == nil {
-		handle.Size = inf.Size()
+		options.Handle.Size = inf.Size()
 	}
 
-	handle.UnixFD = uint64(f.Fd())
+	options.Handle.UnixFD = uint64(f.Fd())
 	if !fc.offloadIO {
-		handle.Flags.Set(handlemap.HandleFlagCached)
+		options.Handle.Flags.Set(handlemap.HandleFlagCached)
 	}
 
-	log.Info("FileCache::OpenFile : file=%s, fd=%d", options.Name, f.Fd())
-	handle.SetFileObject(f)
+	log.Info("FileCache::DownloadFile : file=%s, fd=%d", options.Name, f.Fd())
+	options.Handle.SetFileObject(f)
+
+	//set boolean in isDownloadNeeded value to signal that the file has been downloaded
+	noDownloadNeeded := struct{ downloadNeeded bool }{downloadNeeded: false}
+	options.Handle.SetValue("isDownloadNeeded", noDownloadNeeded)
+
+	return options.Handle, nil
+}
+
+// OpenFile: Makes the file available in the local cache for further file operations.
+func (fc *FileCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
+	log.Trace("FileCache::OpenFile : name=%s, flags=%d, mode=%s", options.Name, options.Flags, options.Mode)
+
+	_, err := fc.NextComponent().GetAttr(internal.GetAttrOptions{Name: options.Name})
+
+	// return err in case of authorization permission mismatch
+	if err != nil && err == syscall.EACCES {
+		return nil, err
+	}
+
+	fileCacheStatsCollector.UpdateStats(stats_manager.Increment, dlFiles, (int64)(1))
```

Comment:
The original code has two different versions of this call, with `dlFiles` when it downloads the file, and `cacheServed` when it doesn't.
So this call probably needs to move to `DownloadFile`.